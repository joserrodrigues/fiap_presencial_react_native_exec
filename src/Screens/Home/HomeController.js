import React from 'react';
import HomeView from './HomeView';
import { registerRootComponent } from 'expo';

const HomeController = () => {
    return (
        <HomeView />
    );
};

export default registerRootComponent(HomeController);