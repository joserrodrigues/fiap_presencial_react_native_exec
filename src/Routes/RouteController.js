import 'react-native-gesture-handler';
import * as React from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';
import { createDrawerNavigator } from '@react-navigation/drawer';
import { registerRootComponent } from 'expo';
import { Icon } from 'react-native-elements';

import Colors from '../Utils/Constants/Colors';
import HomeController from '../Screens/Home/HomeController';
import DetailController from '../Screens/Detail/DetailController';
import MyPositionController from '../Screens/MyPosition/MyPositionController';
import MyInfoController from '../Screens/MyInfo/MyInfoController';
import LoginController from '../Screens/Login/LoginController';

import store from '../store/stores';
import { Provider } from 'react-redux';
import { useDispatch, useSelector } from 'react-redux';
import SideMenu from '../Components/SideMenu/SideMenu';
import { useGetStorageItem } from '../Services/Storage/StorageServices';
import { stopLoadUserInfo, logout } from '../store/modules/login/actions';

const Stack = createStackNavigator();
const Drawer = createDrawerNavigator();

function MainRouteController() {

    const dispatch = useDispatch();
    const getAsyncInfo = async () => {
        useGetStorageItem("userInfo").then((info) => {
            dispatch(stopLoadUserInfo(info));
        })
    }

    const onLogout = () => {        
        dispatch(logout());
    }

    let screenOptions = {
        headerShown: true,
        headerStyle: {
            backgroundColor: Colors.HeaderBackgroundColor,
        },
        headerTintColor: Colors.HeaderTintColor,
        headerLayoutPreset: 'center',
    };

    const StackHome = () => {
        return (
            <Stack.Navigator>
                <Stack.Screen name="Home" component={HomeController}
                    options={screenOptions} />
                <Stack.Screen name="Details" component={DetailController}
                    options={screenOptions} />
            </Stack.Navigator>
        );
    }

    const StackMyInfo = () => {
        return (
            <Stack.Navigator>
                <Stack.Screen name="MyInfo" component={MyInfoController}
                    options={{ ...screenOptions, title: "Minhas Informações" }} />
            </Stack.Navigator>
        );
    }

    const StackMyPosition = () => {
        return (
            <Stack.Navigator>
                <Stack.Screen name="MyPosition" component={MyPositionController}
                    options={{ ...screenOptions, title: "Minha Posição" }} />
            </Stack.Navigator>
        );
    }
    
    const userInfo = useSelector((state) => state.loginSaga.userInfo);
    let hasToken = false;
    if (userInfo !== undefined && userInfo !== null) {
        hasToken = true;
    } else {
        getAsyncInfo();
    }

    if (!hasToken) {
        return (
            <NavigationContainer>
                <Stack.Navigator>
                    <Stack.Screen name="MyPosition" component={LoginController}
                        options={{ headerShown: false }} />
                </Stack.Navigator>
            </NavigationContainer>
        )
    } else {
        return (
            <NavigationContainer>
                <Drawer.Navigator
                    initialRouteName="Main"
                    // openByDefault="open"
                    drawerContent={(props) =>                        
                        <SideMenu userInfo={userInfo} logout={onLogout} {...props} />
                    }
                    drawerStyle={{
                        backgroundColor: Colors.Red,
                        width: 240,
                    }}
                    drawerContentOptions={{
                        activeTintColor: Colors.White,
                        inactiveTintColor: Colors.DarkGrey,
                    }}
                    headerShown={false}
                    >
                    <Drawer.Screen name="Main" component={StackHome}             
                        options={{
                            drawerIcon: ({ color, size }) => {
                                return (
                                    <Icon
                                        name='home'
                                        type='font-awesome'
                                        size={size}
                                        color={color}
                                    />
                                )
                            }
                            
                        }} 
                    />
                    <Drawer.Screen name="MyInfo" component={StackMyInfo}
                        options={{
                            drawerLabel: 'Minha Informação',
                            drawerIcon: ({ color, size }) => (
                                <Icon
                                    name='info-circle'
                                    type='font-awesome-5'
                                    size={size}
                                    color={color}
                                />
                            ) }} />
                    <Drawer.Screen name="MyPosition" component={StackMyPosition}
                        options={{
                            drawerLabel: 'Minha Posição',
                            drawerIcon: ({ color, size }) => (
                                <Icon
                                    name='location'
                                    type='ionicon'
                                    size={size}
                                    color={color}
                                />
                            ) }} />
                </Drawer.Navigator>
            </NavigationContainer>
        );    
    }
}


function RouteController() {
    return (
        <Provider store={store}>
            <MainRouteController />
        </Provider>
    )
}

export default registerRootComponent(RouteController);