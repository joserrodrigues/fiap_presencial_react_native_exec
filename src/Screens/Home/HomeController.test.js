import React from 'react';
import renderer from 'react-test-renderer';
import api from '../../Services/APIs/Common/api';
import MockAdapter from "axios-mock-adapter";
import { NavigationContainer } from '@react-navigation/native';

import { render, fireEvent, waitForElementToBeRemoved } from '@testing-library/react-native';
import { useNavigation } from '@react-navigation/native';
import HomeController from './HomeController';
import { StackHome } from '../../Routes/RouteController';

const fakeList = {
    persons: [
        {
            id: 0,
            firstName: 'Geny',
            lastName: 'Batista',
            jobTitle: 'Assistente Configuração Interno',
            jobArea: 'Diretivas',
            jobDescriptor: 'Internacional',
            jobType: 'Associado',
            address: '16659 Eliana Marginal',
            zipCode: '96749-400',
            city: 'Nova Taiane',
            state: 'AL',
            coutry: 'Algéria',
            phone: '(39) 0220-0167',
            CPF: '30532433424',
            latitude: '27.4159',
            longitude: '-115.0988',
            image: 'https://random.imagecdn.app/500/200'
        }
    ]
};
jest.useRealTimers();
jest.setTimeout(30000);
beforeAll(() => {
    console.log(" IS MOCKING = " + global.__isMocking__);
    if (global.__isMocking__) {
        mock = new MockAdapter(api);
    }
});

afterEach(() => {
    if (global.__isMocking__) {
        mock.reset();
    }
});

describe('<HomeController />', () => {
    it('Check Initial Case', async () => {
        if (global.__isMocking__) {
            mock.onGet('/persons/getPersons').reply(200, fakeList);
        }
        
        const navigation = useNavigation();
        const tree = renderer.create(
            <HomeController navigation={navigation} />
        ).toJSON();
        expect(tree.children.length).toBe(3);
        expect(tree).toMatchSnapshot();
    });

    it('Mocking Info', async () => {
        if (global.__isMocking__) {
            mock.onGet('/persons/getPersons').reply(200, fakeList);
        }

        const navigation = useNavigation();
        const { getByTestId } = render(<HomeController navigation={navigation} />);

        const element = await waitForElementToBeRemoved(() => getByTestId('activityLoading'),
            { timeout: 1000500, interval: 10000 });
        expect(element).toBeTruthy();
        const flatList = getByTestId('flatListHome');
        console.log("# of regiters = " + flatList._fiber.stateNode.props.data.length);
        if (global.__isMocking__) {
            expect(flatList._fiber.stateNode.props.data.length).toBeLessThanOrEqual(50);
        } else {
            expect(flatList._fiber.stateNode.props.data.length).toBeGreaterThanOrEqual(50);
        }
    });


    it('Click Home Info', async () => {
        if (global.__isMocking__) {
            mock.onGet('/persons/getPersons').reply(200, fakeList);
        }

        const { getByTestId, getByText } = render(
            <NavigationContainer>
                <StackHome />
            </NavigationContainer>
        );

        const element = await waitForElementToBeRemoved(() => getByTestId('activityLoading'),
            { timeout: 1000500, interval: 10000 });
        expect(element).toBeTruthy();
        const flatList = getByTestId('flatListHome');
        console.log("# of regiters = " + flatList._fiber.stateNode.props.data.length);
        if (global.__isMocking__) {
            expect(flatList._fiber.stateNode.props.data.length).toBeLessThanOrEqual(50);
        } else {
            expect(flatList._fiber.stateNode.props.data.length).toBeGreaterThanOrEqual(50);
        }

        const firstItem = flatList._fiber.stateNode.props.data[0];
        console.log(firstItem.CPF.toString());

        let firstButton = getByTestId("button" + firstItem.CPF.toString());
        expect(firstButton).toBeTruthy();
        fireEvent.press(firstButton);

        const newScreen = await getByText('Detalhe');
        expect(newScreen).toBeTruthy();
    });
});