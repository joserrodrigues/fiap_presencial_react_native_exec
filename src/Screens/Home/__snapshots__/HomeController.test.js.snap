// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeController /> Check Initial Case 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "flex-start",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderBottomColor": "#000",
        "borderBottomWidth": 1,
        "borderTopColor": "#000",
        "borderTopWidth": 1,
        "marginBottom": 10,
        "padding": 8,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 0,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#eeeeee",
            "borderBottomWidth": 0,
            "borderColor": "#86939e",
            "borderRadius": 20,
            "flexDirection": "row",
            "minHeight": 30,
            "overflow": "hidden",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginLeft": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
        <TextInput
          allowFontScaling={true}
          editable={true}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Digite Aqui..."
          placeholderTextColor="#86939e"
          platform="default"
          rejectResponderTermination={true}
          style={
            Object {
              "color": "#86939e",
              "flex": 1,
              "fontSize": 18,
              "marginLeft": 10,
              "minHeight": 40,
            }
          }
          testID="searchInput"
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginRight": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      />
    </View>
  </View>
  <ActivityIndicator
    animating={true}
    color="#fa5757"
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "marginTop": 30,
      }
    }
    testID="activityLoading"
  />
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="flatListHome"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</RCTSafeAreaView>
`;
